src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLambdaParameterShouldBeUnnamedCheck.java

```diff
@@ -0,0 +1,284 @@
+ * </ul>
+ *
+ * @since 10.18.0
+ *
+ */
+@FileStatefulCheck
+public class UnusedLambdaParameterShouldBeUnnamedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_LAMBDA_PARAMETER = "unused.lambda.parameter";
+
+    /**
+     * Invalid parents of the lambda parameter identifier.
+     */
+    private static final int[] INVALID_LAMBDA_PARAM_IDENT_PARENTS = {
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * Keeps track of the lambda parameters in a block.
+     */
+    private final Deque<LambdaParameterDetails> lambdaParameters = new ArrayDeque<>();
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.LAMBDA,
+            TokenTypes.IDENT,
+        };
+    }
+
+    @Override
+    public void beginTree(DetailAST rootAST) {
+        lambdaParameters.clear();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (ast.getType() == TokenTypes.LAMBDA) {
+            final DetailAST parameters = ast.findFirstToken(TokenTypes.PARAMETERS);
+            if (parameters != null) {
+                // we have multiple lambda parameters
+                TokenUtil.forEachChild(parameters, TokenTypes.PARAMETER_DEF, parameter -> {
+                    final DetailAST identifierAst = parameter.findFirstToken(TokenTypes.IDENT);
+                    final LambdaParameterDetails lambdaParameter =
+                            new LambdaParameterDetails(ast, identifierAst);
+                    lambdaParameters.push(lambdaParameter);
+                });
+            }
+            else if (ast.getChildCount() != 0) {
+                // we are not switch rule and have a single parameter
+                final LambdaParameterDetails lambdaParameter =
+                            new LambdaParameterDetails(ast, ast.findFirstToken(TokenTypes.IDENT));
+                lambdaParameters.push(lambdaParameter);
+            }
+        }
+        else if (isLambdaParameterIdentifierCandidate(ast) && !isLeftHandOfAssignment(ast)) {
+            // we do not count reassignment as usage
+            lambdaParameters.stream()
+                    .filter(parameter -> parameter.getName().equals(ast.getText()))
+                    .findFirst()
+                    .ifPresent(LambdaParameterDetails::registerAsUsed);
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        if (ast.getType() == TokenTypes.LAMBDA) {
+            // when leaving a given lambda, check every parameter associated with it
+            while (!lambdaParameters.isEmpty()
```

Comment:
```suggestion
            while (lambdaParameters.peek() != null
```
This will make checker happy

Reply:
done

Reply:
but still checker fails I have no idea how is that possible. may be a false positive? are we ok to suppress?

Reply:
```suggestion
                        && ast.equals(lambdaParameters.peek().enclosingLambda())) {
```
Would this make checker happy?

Reply:
No. very weird

```Java
 while (lambdaParameters.peek() != null
                        && ast.equals(lambdaParameters.peek().enclosingLambda)) { ... }
```
```XML
  <checkerFrameworkError unstable="false">
    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLambdaParameterShouldBeUnnamedCheck.java</fileName>
    <specifier>dereference.of.nullable</specifier>
    <message>dereference of possibly-null reference lambdaParameters.peek()</message>
    <lineContent>&amp;&amp; ast.equals(lambdaParameters.peek().enclosingLambda)) {</lineContent>
  </checkerFrameworkError>
```

Reply:
I am good to suppress this, we should also open a bug report with checker

Reply:
Leaving this unresolved for the next reviewers to see.

Reply:
https://github.com/typetools/checker-framework/issues/6760